### DISCUSSION TITLE
I think this could become slightly more higher level.

### DISCUSSION THREAD

**Discussion by epigramx** [semuconsulting/pyubx2#112] [Q&A] [answered]
I mean you went all the way to abstracting it with attributes out of the messages it receives and sends and it remains in need of checking the documentation on the protocols in most cases.
What I mean is methods like "just turn off the messages of class X and ID Y"; or "turn on messages if their port type is USB"; it's already great just mentioning it feels it can be more.

**Comment by semuadmin** [comment, marked as answer]
Sorry @epigramx I'm not entirely clear what the issue is here? 

pyubx2 is a UBX parser and generation utility library - it's not an end-user application. If you need higher level functionality - like for example the ability to turn on or off all UBX messages of a certain category - then it's a relatively simple matter to build this yourself using this library. There are plenty of examples in the [/examples](https://github.com/semuconsulting/pyubx2/tree/master/examples) folder (see, for instance, [ubxsetrates.py](https://github.com/semuconsulting/pyubx2/blob/master/examples/ubxsetrates.py)), but if you need specific guidance, raise a query in the [Discussions channel](https://github.com/semuconsulting/pyubx2/discussions) and we'll do our best to assist.

[PyGPSClient](https://github.com/semuconsulting/PyGPSClient) is an example of an end-user application which uses the pyubx2 library, and it does have the functionality you're describing as well as the ability to define custom command strings as 'presets'. See the following links for further information:

https://github.com/semuconsulting/PyGPSClient
https://github.com/semuconsulting/PyGPSClient#userdefined

In addition, [pygnssutils](https://github.com/semuconsulting/pygnssutils#ubxsetrate) offers a set of command line utilities including a `ubxsetrate` function which allows you to set the rate for a given message using a simple terminal command. Again, if you would like to see additional command line functionality, make a suggestion in the [pygnssutils Discussion Channel](https://github.com/semuconsulting/pygnssutils/discussions).
```
➜  ~ ubxsetrate -h
usage: ubxsetrate [-h] [-V] -P PORT [--baudrate {4800,9600,19200,38400,57600,115200,230400,460800}] [--timeout TIMEOUT] --msgClass
                  MSGCLASS [--msgID MSGID] [--rate RATE]

options:
  -h, --help            show this help message and exit
  -V, --version         show program's version number and exit
  -P PORT, --port PORT  Serial port (default: None)
  --baudrate {4800,9600,19200,38400,57600,115200,230400,460800}
                        Serial baud rate (default: 9600)
  --timeout TIMEOUT     Serial timeout in seconds (default: 3.0)
  --msgClass MSGCLASS   Message class from pyubx2.UBX_CLASSES or special values 'allubx', 'minubx', 'allnmea' or 'minnmea'
                        (default: None)
  --msgID MSGID         Message ID from pyubx2.UBX_MSGIDS[1:] (default: None)
  --rate RATE           Message rate per navigation solution (default: 1)

© 2022 SEMU Consulting BSD 3-Clause license - https://github.com/semuconsulting/pygnssutils/
```

**Comment by epigramx** [comment]
> If you need higher level functionality - like for example the ability to turn on or off all UBX messages of a certain category - then it's a relatively simple matter to build this yourself

Yep that's what I'm doing right now. By the way: do you have an example on how to turn on RXM-RAWX on USB? I found out how to do it for the CFG-MSG NMEA messages.

**Reply by semuadmin** [reply]
Assuming your receiver supports the RXM-RAWX message (refer to the device's Integration Manual), you can enable it in exactly the same way you would enable any other UBX message - by sending a CFG-MSG command with the appropriate msgClass, msgID and rate values.

RXM-RAWX = msgClass 0x02, msgId 0x15

Alternatively, for Generation 9+ devices only, you can use the more modern CFG-VALSET command with the appropriate configuration database key(s). `pyubx2` provides a helper method `config_set()` to help format these messages (_NB: there are over [1200 possible database keys](https://github.com/semuconsulting/pyubx2/blob/master/src/pyubx2/ubxtypes_configdb.py), but their string names are on the whole fairly self-explanatory_) e.g.:

```
from serial import Serial
from pyubx2 import UBXMessage, SET, SET_LAYER_RAM, TXN_NONE

# substitute your port and baudrate as required
with Serial("/dev/tty.usbmodem101", 38400, timeout=3) as rcvr:

    # legacy method using CFG-MSG
    msg = UBXMessage(
        "CFG",
        "CFG-MSG",
        SET,
        msgClass=0x02,
        msgID=0x15,
        rateUSB=1,
    )
    rcvr.write(msg.serialize())

    # modern method (for Generation 9+ devices only) using CFG-VALSET
    msg = UBXMessage.config_set(
        SET_LAYER_RAM, TXN_NONE, [("CFG_MSGOUT_UBX_RXM_RAWX_USB", 1)]
    )
    rcvr.write(msg.serialize())

```

Alternatively, using the `ubxsetrate` command line utility (this sets the rate on all available output ports):

```
➜  ~ ubxsetrate -P /dev/tty.usbmodem101 --baudrate 38400 --msgClass 0x02 --msgID 0x15 --rate 1
Opening serial port /dev/tty.usbmodem101 @ 38400 baud ...

Sending configuration message <UBX(CFG-MSG, msgClass=RXM, msgID=RXM-RAWX, rateDDC=1, rateUART1=1, rateUART2=1, rateUSB=1, rateSPI=1, reserved=0)>...

Configuration message(s) sent.
```

**Comment by epigramx** [comment]
Oh that was silly. I was expecting `UBXMessage()` to need `RXM-RAWX` as a 2nd and `RXM` as 1st argument in that case (I guess it's simpler than I thought). Thanks very much I'll look into it.

**Reply by semuadmin** [reply]
> Oh that was silly. I was expecting `UBXMessage()` to need `RXM-RAWX` as a 2nd and `RXM` as 1st argument in that case (I guess it's simpler than I thought).

Exactly - you're not creating an RXM-RAWX message; you're creating a CFG-MSG message to enable RXM-RAWX output.
